rails_programming/turbo/stimulus_js.md

```diff
@@ -0,0 +1,287 @@
+# Stimulus JS
+
+### Introduction
+
+Remember JavaScript? What did we learn that for? Well, sometimes you want a bit more interactivity in your browser, 
+like hiding/showing a part of your page. Making a round trip to your server is impractical. That is where JavaScript
+comes in. So Rails suggest using Stimulus to augment your html. Stimulus gives you a way of creating and using reusable 
+controllers that give you that extra little bit of interactivity.
+
+### Learning Outcomes
+
+*Look through these now and then use them to test yourself after doing the assignment*
+
+After going through this lesson you should be able to
+
+* explain how to attach Stimulus controllers to your html
+* how to use targets instead of query selectors
+* how to use action attributes instead of event listeners
+* where to keep state
+* how to use lifecycle callbacks
+* how to make your controllers reusable
+* how to make your controllers customizable with attributes
+
+### The idea
+
+Stimulus is a modest framework. It leaves things mostly up to you, but gives you a consistent way of doing things.
+It uses html `data-attributes` to attach and configure behavior on your html. Let us look an an html example
+from the handbook:
+
+```
+<div data-controller="clipboard">
+  PIN: <input data-clipboard-target="source" type="text" value="3737" readonly>
+  <button data-action="click->clipboard#copy">Copy to Clipboard</button>
+</div>
+```
+
+Read carefully through the example, and pay special attention to the data attributes! Can you guess what it does?
+
+This example is using three data attributes: `data-controller`, `data-clipboard-target`, `data-action`. So what is their
+purpose?
+
+* **data-controller** is set to `clipboard`. Which means: Hey Stimulus, we want to enhance this html with some clipboard
+behavior
+* **data-clipboard-target** is set to `source`. Which says: this is the the `source` we are gonna use for our clipboard
+controller
+* **data-action** is set to `click->clipboard#copy`. Which means: On a click invoke the `copy`action the our clipboard
+controller
+
+So together these three attributes enable the following behavior: When a user `click`s on the button, use the function
+`copy` of Stimulus controller `clipboard` with the input as a `source`. The yet-to-be-seen JavaScript controller will
+then take care of the rest and will copy the value of the source target to the users clipboard.
+
+Wasn't this supposed to be a JavaScript lesson? Yes, but the point is that Stimulus is a HTML first framework. It is
+generally not concerned with rendering html, but rather with attaching behavior to it with sprinkles of JavaScript.
+Trough the consistent use of those data attributes we can read the html and can discover how the html was enhanced,
+we don't have to guess which even listener might have been have been attached to an element or which element
+serves in certain role.
```

Comment:
I think we need to adapt this slightly to be more readable.

Reply:
Good point, I put in an example to make the point clearer:
```markdown

Wasn't this supposed to be a JavaScript lesson? Yes, but the point is that Stimulus is a HTML first framework. 
It generally does not render html, instead it attaches behavior to your existing html with sprinkles of JavaScript.
Trough the consistent use of some data attributes we can read the html and can see where the html is enhanced.

Look this html without the special data-attributes Stimulus uses:
```
```html
<input id="pin-code value="1337" readonly>
<button id="pin-button">Copy to Clipboard</button>
```

```markdown
Probably something is supposed to happen, when you click that button. However you can't tell from the html
alone how things are wired up. You have to search for some JavaScript that handles this behavior.
```

Reply:
Thanks. Just one small typo of `trough` instead of `through`

Reply:
Thx, also adding some checks now